Add support for HLS and WebRTC Octoprint cameras

vexofp opened the pull request:
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

Octoprint has for several years now supported HLS and WebRTC camera streams, in addition to legacy MJPEG streams. The Home Assistant integration for Octoprint, however, currently only creates `MjpegCamera`. This PR add supports for detecting these additional stream types and wiring them up to `GenericCamera`.

It adds tests that verify the correct camera type is created, based on the URL, as well as tests around WebRTC offer signaling requests via the websocket. With all the new tests, things were getting a bit repetitive, so I refactored everything to use some new fixtures for camera settings test data instead of having it pasted everywhere. (I'm still learning my way around all of HA's testing framework, so hopefully this is close to the mark. 🤞)

It also bumps the required version of `pyoctoprintapi` to `0.1.14` ([changelog](https://github.com/rfleming71/pyoctoprintapi/releases/tag/v0.1.14)), which is necessary to retrieve the WebRTC URLs correctly (https://github.com/rfleming71/pyoctoprintapi/issues/12). This required a few updates to the test job data to set the `["job"]["file"]`, as this field is required by the new version.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

vexofp committed:
Add support for HLS and WebRTC Octoprint cameras

MartinHjelmare commented on homeassistant/components/octoprint/camera.py:
```diff
@@ -1,17 +1,37 @@
+    CONF_LIMIT_REFETCH_TO_URL_CHANGE,
+    CONF_STILL_IMAGE_URL,
+    CONF_STREAM_SOURCE,
+    GenericCamera,
```
The GenericCamera isn't written or meant to be used as a helper for other integrations. If we want to allow that it needs to be refactored first to have a proper interface with keyword only arguments instead of passing a dict with config items. The class also needs to be exported from the generic package via `__all__` so it's clear that other integrations can import it.

We did the same changes for `MjpegCamera` earlier.

I'll check with the core team of we think it's good to be able to reuse the generic camera.

vexofp replied:
Okay. We can take a look at what was done with `MjpegCamera` and try to make a PR to see what that looks like for `GenericCamera`.

MartinHjelmare replied:
The conclusion from the core team is it's ok to use the generic camera integration but the camera class should be refactored or split so that it gets an interface that other integrations can use, per my comment above.

jeeftor committed:
Weatherflow_cloud backing lib bump (#112262)

Backing lib bump

synesthesiam committed:
Bump intents to 2024.3.6 (#112515)

jeeftor committed:
Bump weatherflow4py to 0.1.14 (#112554)

adding missing rain states

emontnemery committed:
Restore the juicenet integration (#112578)

bramkragten committed:
Update frontend to 20240307.0 (#112620)

emontnemery committed:
Make hass-nabucasa a core requirement (#112623)

janiversen committed:
modbus scan_interval: 0 is correct configuration (#112619)

janiversen committed:
Bump pymodbus to v3.6.5 (#112629)

Kane610 committed:
Bump axis to v52 (#112632)

* Bump axis to v51

* Bump to v52

bachya committed:
Bump `aionotion` to 2024.03.0 (#112675)

Commit:
Fix incorrect filtering of unsupported locales in bring-api (#112589)

bieniu committed:
Downgrade `pysnmp-lextudio` to version `5.0.34` (#112696)

Downgrade pysnmp-lextudio to version 5.0.34

Co-authored-by: Maciej Bieniek <478555+bieniu@users.noreply.github.com>

janiversen committed:
Allow duplicate names in different modbus entities (#112701)

Allow duplicate names in different entities.

puddly committed:
Disable updating ZHA coordinator path from discovery info (#112415)

* Never update the device path from config flows

* Bring coverage up to 100%

* Update tests/components/zha/test_config_flow.py

Co-authored-by: TheJulianJES <TheJulianJES@users.noreply.github.com>

---------

Co-authored-by: TheJulianJES <TheJulianJES@users.noreply.github.com>

catsmanac committed:
Bump pyenphase to 1.19.2 (#112747)

alinbalutoiu committed:
Fix tado climate service (#112686)

bdraco committed:
Guard against db corruption when renaming entities (#112718)

bachya committed:
Make sure Notion saves new refresh token upon startup (#112676)

* Make sure Notion saves new refresh token upon startup

* Code review

* Typing

* Smoother syntax

* Fix tests

* Fix tests for real

allenporter committed:
Fix local calendar handling of empty recurrence ids (#112745)

* Fix handling of empty recurrence ids

* Revert logging changes

thecode committed:
Bump bthome-ble to 3.7.0 (#112783)

janiversen committed:
Issue warning modbus configuration when modbus configuration is empty (#112618)

lextm committed:
Upgrade `pysnmp-lextudio` to version `6.0.9` (#112795)

bdraco committed:
Fix MJPEG fallback when still image URL is missing with basic auth (#112861)

* Fix MJPEG fallback when still image URL is missing with basic auth

I picked up an old DCS-930L (circa 2010) camera to test with
to fix #94877

* Fix MJPEG fallback when still image URL is missing with basic auth

I picked up an old DCS-930L (circa 2010) camera to test with
to fix #94877

* Fix MJPEG fallback when still image URL is missing with basic auth

I picked up an old DCS-930L (circa 2010) camera to test with
to fix #94877

* Fix MJPEG fallback when still image URL is missing with basic auth

I picked up an old DCS-930L (circa 2010) camera to test with
to fix #94877

clssn committed:
Update the numato-gpio dependency of the numato integration to v0.12.0 (#112272)

* Update the numato-gpio dependency of the numato integration to v0.12.0

* Augment numato integration manifest with integration_type

Fulfills a requirement in the PR checklist.

bachya committed:
Streamline Notion config entry updates (refresh token and user ID) (#112832)

dknowles2 committed:
Add missing translation placeholder in Hydrawise (#113007)

Add missing translation placeholder

Thomas55555 committed:
Bump aioautomower to 2024.3.0 (#112627)

Fix error in Husqvarna automower in Zones dataclass

allenporter committed:
Bump ical to 7.0.1 and always use home assistant timezone for local todo dtstart (#113034)

Kane610 committed:
Bump axis to v53 (#113019)

Ernst79 committed:
Bump bthome-ble to 3.8.0 (#113008)

Bump bthome-ble

FlyingFeng2021 committed:
Bump boschshcpy to 0.2.82 (#112890)

marcelveldt committed:
Fix for controlling Hue switch entities (#113064)

folke committed:
Fix hvac_mode for viessmann devices with heatingCooling mode (#113054)

LukasdeBoer committed:
Bump rova to 0.4.1 (#113066)

alistair23 committed:
components/gardena_bluetooth: Improve avaliability reliability (#113056)

* components/gardena_bluetooth: Improve avaliability reliability

The available() function incorrectly returns false even though the device
is accessible.

The last_update_success property should correctly indicate if the device
isn't contactable, so we don't need to call async_address_present().
This is confirmed by seeing that no other users are calling
async_address_present() in the available() function.

This commit removes the call to async_address_present() to help fix the
sensor being unavailable when using a ESPHome BLE proxy.

Signed-off-by: Alistair Francis <alistair@alistair23.me>

---------

Signed-off-by: Alistair Francis <alistair@alistair23.me>
Co-authored-by: Joakim Plate <elupus@ecce.se>

marcelveldt committed:
Fix colormode attribute on grouped Hue light (#113071)

RunC0deRun committed:
Fix optional Jellyfin RunTimeTicks (#108254)

bdraco committed:
Fix failing google diagnostics test (#113095)

mikewoudenberg committed:
Update Loqed helper for more logging (#112646)

Updates Loqed helper for more logging

bdraco committed:
Bump aiodhcpwatcher to 0.8.1 (#113096)

Kane610 committed:
Bump axis to v54 (#113091)

Co-authored-by: J. Nick Koston <nick@koston.org>

zweckj committed:
bump pytedee_async to 0.2.16 (#113135)

jbouwh committed:
Fix google_asssistant sensor state reporting (#112838)

* Fix post google_assistant sensor values as float not string

* Fix aqi reporting and improve tests

* Fix _air_quality_description_for_aqi and  test

zweckj committed:
Include pytedee_async logger in tedee integration (#112590)

add pytedee logger

StevenLooman committed:
Improve discovering upnp/igd device by always using the SSDP-discovery for the Unique Device Name (#111487)

* Always use the UDN found in the SSDP discovery, instead of the device description

* Ensure existing DeviceEntries are still matched

bieniu committed:
Fix availability for GIOS index sensors (#113021)

* Fix availability for index sensors

* Improve test_availability()

---------

Co-authored-by: Maciej Bieniek <478555+bieniu@users.noreply.github.com>

cottsay committed:
Fix some handle leaks in rainforest_raven (#113035)

There were leaks when
* The component was shutdown
* There was a timeout during the initial device opening

Additionally, the device was not closed/reopened when there was a
timeout reading regular data.

synesthesiam committed:
Bump intents to 2023.3.12 (#113160)

Bump intents

Thomas55555 committed:
Bump aioautomower to 2024.3.2 (#113162)

mrchi committed:
Bump openwrt-luci-rpc version to 1.1.17 (#112796)

jeeftor committed:
Bump weatherflow4py to 0.1.17 (#112661)

OnFreund committed:
Use friendly name for camera media source (#110882)

mattmccormack committed:
Add auto fan mode icon (#110185)

LaStrada committed:
Bump airthings_ble to 0.7.1 (#113172)

Co-authored-by: Ståle Storø Hauknes <LaStrada@users.noreply.github.com>

ctalkington committed:
Bump rokuecp to 0.19.2 (#113198)

agners committed:
Add message from Bad Request errors to HassioAPIError (#113144)

Co-authored-by: Mike Degatano <michael.degatano@gmail.com>

ctalkington committed:
Bump pyipp to 0.15.0 (#113204)

update pyipp to 0.15.0

bieniu committed:
Bump `brother` library to version `4.0.2` (#113235)

Co-authored-by: Maciej Bieniek <478555+bieniu@users.noreply.github.com>

FieldofClay committed:
Ignore AussieBroadband services that don't support usage information (#110253)

frenck committed:
Bump version to 2024.3.1

frenck committed:
Hotfix import error in ZHA for 2024.3.1 patch release (#113250)

jonnybergdahl committed:
Fix Twitch auth token refresh (#112833)

* Fix for expired token

* Add auth token refresh.

* Eliminate extra auth call

* Fixed mock client

---------

Co-authored-by: Jonny Bergdahl <bergdahl@users.noreply.github.com>

ctalkington committed:
Add diagnostics for IPP (#113205)

Thomas55555 committed:
Add loggers to Husqvarna Automower (#113381)

mdegat01 committed:
Apply suggestion failures fail supervisor repair (#113372)

mdegat01 committed:
Supervisor issues update retries on failure (#113373)

frenck committed:
2024.3.1 (#113249)

Thomas55555 committed:
Bump aioautomower to 2024.3.3 (#113430)

Commit:
Check for EA release channel for UniFi Protect (#113432)

Co-authored-by: J. Nick Koston <nick@koston.org>

bdraco committed:
Bump aiodhcpwatcher to 0.8.2 (#113466)

lextm committed:
Bump `pysnmp-lextudio` to version `6.0.11` (#113463)

Kane610 committed:
Bump axis to v55 (#113479)

OnFreund committed:
Revert setting communication delay in Risco init (#113497)

dgomes committed:
Bump croniter to 2.0.2 (#113494)

OnFreund committed:
Bump pyrisco to 0.5.10 (#113505)

joostlek committed:
Fix Airthings BLE illuminance sensor name (#113560)

thecode committed:
Ignore Shelly block update with cfgChanged None (#113587)

bieniu committed:
Catch `TimeoutError` in `Brother` config flow (#113593)

* Catch TimeoutError in Brother config flow

* Update tests

* Remove unnecessary parentheses

---------

Co-authored-by: Maciej Bieniek <478555+bieniu@users.noreply.github.com>

Kane610 committed:
Bump axis to v56 (#113608)

allenporter committed:
Bump ical to 7.0.3 to fix local-todo persisted with invalid DTSTART values (#113526)

Commit:
Bump pyunifiprotect to 5.0.1 (#113630)

emontnemery committed:
Fix missing context when running script from template entity (#113523)

Co-authored-by: J. Nick Koston <nick@koston.org>

Commit:
Bump pyunifiprotect to 5.0.2 (#113651)

bieniu committed:
Add removal condition to Shelly battery sensor (#113703)

Co-authored-by: Maciej Bieniek <478555+bieniu@users.noreply.github.com>

epenet committed:
Fix unknown values in onewire (#113731)

* Fix unknown values in onewire

* Update tests

janiversen committed:
Bump pymodbus v3.6.6 (#113796)

emontnemery committed:
Fix startup race in cast (#113843)

emontnemery committed:
Bump pychromecast to 14.0.1 (#113841)

ludeeus committed:
Redact the area of traccar server geofences (#113861)

cottsay committed:
Bump aioraven to 0.5.2 (#113714)

erwindouna committed:
Tado fix water heater (#113464)

Co-authored-by: Joostlek <joostlek@outlook.com>

emontnemery committed:
Catch API errors in cast media_player service handlers (#113839)

* Catch API errors in cast media_player service handlers

* Remove left over debug code

* Fix wrapping of coroutine function with api_error

dependabot[bot] committed:
Bump home-assistant/builder from 2024.01.0 to 2024.03.5 (#113887)

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

zweckj committed:
Bump pytedee_async to 0.2.17 (#113933)

Kane610 committed:
Bump axis to v57 (#113952)

balloob committed:
Bump version to 2024.3.2

balloob committed:
2024.3.2 (#113973)

* Streamline Notion config entry updates (refresh token and user ID) (#112832)

* Bump aioautomower to 2024.3.2 (#113162)

* Bump aioautomower to 2024.3.3 (#113430)

* Check for EA release channel for UniFi Protect (#113432)

Co-authored-by: J. Nick Koston <nick@koston.org>

* Bump `pysnmp-lextudio` to version `6.0.11` (#113463)

* Tado fix water heater (#113464)

Co-authored-by: Joostlek <joostlek@outlook.com>

* Bump aiodhcpwatcher to 0.8.2 (#113466)

* Bump axis to v55 (#113479)

* Bump croniter to 2.0.2 (#113494)

* Revert setting communication delay in Risco init (#113497)

* Bump pyrisco to 0.5.10 (#113505)

* Fix missing context when running script from template entity (#113523)

Co-authored-by: J. Nick Koston <nick@koston.org>

* Bump ical to 7.0.3 to fix local-todo persisted with invalid DTSTART values (#113526)

* Fix Airthings BLE illuminance sensor name (#113560)

* Ignore Shelly block update with cfgChanged None (#113587)

* Catch `TimeoutError` in `Brother` config flow (#113593)

* Catch TimeoutError in Brother config flow

* Update tests

* Remove unnecessary parentheses

---------

Co-authored-by: Maciej Bieniek <478555+bieniu@users.noreply.github.com>

* Bump axis to v56 (#113608)

* Bump pyunifiprotect to 5.0.1 (#113630)

* Bump pyunifiprotect to 5.0.2 (#113651)

* Add removal condition to Shelly battery sensor (#113703)

Co-authored-by: Maciej Bieniek <478555+bieniu@users.noreply.github.com>

* Bump aioraven to 0.5.2 (#113714)

* Fix unknown values in onewire (#113731)

* Fix unknown values in onewire

* Update tests

* Bump pymodbus v3.6.6 (#113796)

* Catch API errors in cast media_player service handlers (#113839)

* Catch API errors in cast media_player service handlers

* Remove left over debug code

* Fix wrapping of coroutine function with api_error

* Bump pychromecast to 14.0.1 (#113841)

* Fix startup race in cast (#113843)

* Redact the area of traccar server geofences (#113861)

* Bump pytedee_async to 0.2.17 (#113933)

* Bump axis to v57 (#113952)

* Bump version to 2024.3.2

---------

Co-authored-by: Aaron Bach <bachya1208@gmail.com>
Co-authored-by: Thomas55555 <59625598+Thomas55555@users.noreply.github.com>
Co-authored-by: Christopher Bailey <cbailey@mort.is>
Co-authored-by: J. Nick Koston <nick@koston.org>
Co-authored-by: Lex Li <425130+lextm@users.noreply.github.com>
Co-authored-by: Erwin Douna <e.douna@gmail.com>
Co-authored-by: Joostlek <joostlek@outlook.com>
Co-authored-by: Robert Svensson <Kane610@users.noreply.github.com>
Co-authored-by: Diogo Gomes <diogogomes@gmail.com>
Co-authored-by: On Freund <onfreund@gmail.com>
Co-authored-by: Erik Montnemery <erik@montnemery.com>
Co-authored-by: Allen Porter <allen@thebends.org>
Co-authored-by: Shay Levy <levyshay1@gmail.com>
Co-authored-by: Maciej Bieniek <bieniu@users.noreply.github.com>
Co-authored-by: Maciej Bieniek <478555+bieniu@users.noreply.github.com>
Co-authored-by: Scott K Logan <logans@cottsay.net>
Co-authored-by: epenet <6771947+epenet@users.noreply.github.com>
Co-authored-by: jan iversen <jancasacondor@gmail.com>
Co-authored-by: Joakim Sørensen <ludeeus@ludeeus.dev>
Co-authored-by: Josef Zweck <24647999+zweckj@users.noreply.github.com>

balloob committed:
2024.3.2 - fix (#113978)

frenck committed:
Update cosign to 2.2.3 (#113996)

frenck committed:
Bump version to 2024.3.3

Kane610 committed:
Bump axis to v58 (#114008)

frenck committed:
2024.3.3 (#114005)

frenck commented:
Because there hasn't been any activity on this PR for quite some time now, I've decided to close it for being stale.

Feel free to re-open this PR when you are ready to pick up work on it again 👍

../Frenck
